File: macos/QMK Toolbox/USB/Bootloader/AtmelDFUDevice.m
#import "AtmelDFUDevice.h"

@implementation AtmelDFUDevice

- (id)initWithUSBDevice:(USBDevice *)usbDevice {
    if (self = [super initWithUSBDevice:usbDevice]) {
        if ([self revisionBCD] == 0x0936) {
            self.name = @"QMK DFU";
            self.type = BootloaderTypeQMKDFU;
        } else {
            self.name = @"Atmel DFU";
            self.type = BootloaderTypeAtmelDFU;
        }
        self.eepromFlashable = YES;
        self.resettable = YES;
    }
    return self;
}

-(void)flashWithMCU:(NSString *)mcu file:(NSString *)file {
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"erase", @"--force"]];
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"flash", @"--force", file]];
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"reset"]];
}

-(void)flashEEPROMWithMCU:(NSString *)mcu file:(NSString *)file {
    if (self.type == BootloaderTypeAtmelDFU) {
        [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"erase", @"--force"]];
    }

    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"flash", @"--force", @"--suppress-validation", @"--eeprom", file]];

    if (self.type == BootloaderTypeAtmelDFU) {
        [self printMessage:@"Please reflash device with firmware now" withType:MessageTypeBootloader];
    }
}

-(void)resetWithMCU:(NSString *)mcu {
    [self runProcess:@"dfu-programmer" withArgs:@[mcu, @"reset"]];
}

@end

File: macos/QMK Toolbox/USB/Bootloader/AtmelDFUDevice.swift
import Foundation

class AtmelDFUDevice: BootloaderDevice {
    override init(usbDevice: USBDevice) {
        super.init(usbDevice: usbDevice)
        if revisionBCD == 0x0936 {
            name = "QMK DFU"
            type = .qmkDfu
        } else {
            name = "Atmel DFU"
            type = .atmelDfu
        }
        eepromFlashable = true
        resettable = true
    }

    override func flash(_ mcu: String, file: String) {
        runProcess("dfu-programmer", args: [mcu, "erase", "--force"])
        runProcess("dfu-programmer", args: [mcu, "flash", "--force", file])
        runProcess("dfu-programmer", args: [mcu, "reset"])
    }

    override func flashEEPROM(_ mcu: String, file: String) {
        if type == .atmelDfu {
            runProcess("dfu-programmer", args: [mcu, "erase", "--force"])
        }
        runProcess("dfu-programmer", args: [mcu, "flash", "--force", "--suppress-validation", "--eeprom", file])
        if type == .atmelDfu {
            print(message: "Please reflash device with firmware now", type: .bootloader)
        }
    }

    override func reset(_ mcu: String) {
        runProcess("dfu-programmer", args: [mcu, "reset"])
    }
}
